TFTRT User provided INT8 quantization scales

trevor-m opened the pull request:
TF-TRT now supports the following quantization nodes:
- QuantizeAndDequantizeV2
- QuantizeAndDequantizeV3
- FakeQuantWithMinMaxVars
- FakeQuantWithMinMaxArgs

When these nodes are converted:
1. Their quantization ranges are extracted and stored
2. The nodes are removed
3. The ranges are applied to the relevant tensors

This enables a path for TF-TRT to deploy models trained with quantization in the loop, for example those trained with tf.contrib.quantize.

trt.create_inference_graph() has a new boolean argument, `use_calibration`.
If we are in INT8 mode and use_calibration=True, create_inference_graph will return a calibration graph just like it did previously.
The calibrator will not override ranges provided via quantization nodes.
If we are in INT8 mode and use_calibration=False, a warning will be issued for every tensor which does not have a calibration range. Since TRT may fuse some operations, users may not always need to provide a value for these tensors. If a tensor that TRT needs is missing, the conversion will fail.

This PR also adds support for Relu6 nodes, and fixes a bug with not renaming tensors.

Current Issues/Considerations:
- If a model was trained with quantization nodes in places where TRT will not quantize (i.e. due to op fusion), then accuracy may drop dramatically. The solution is to figure out which ops will be fused by TRT and avoid placing quantization nodes between those ops. Another option is to place clip ops with each quantization node, so that if the tensor is not quantized it will still be clipped to be in the expected range - this will impact performace negatively.
- TRT does not have documentation for op fusion yet.
- TRT only supports symmetric quantization. Ranges are converted to symmetric ranges using `max(abs(min_range), abs(max_range))`.
- Since the tensors which are inputs and outputs to the TRTEngine are renamed to "TensorRTInputPH_{X}", "TensorRTOutputPH_{X}", if these tensors are missing ranges the warnings will not make sense to users.
- MatMul and BiasAdd are not fused, requiring users to provide a range between these ops.

trevor-m committed:
Allows user to specify int8 scales via TF nodes.

Basic working example - all tensors have range of -5 to 5 [skip ci]

Range of -0.1 to 0.1 seems to work decently [skip ci
]

Load quantization numbers from txt file. Switching between prequant and calib now controlled by two bools (in convert_node.h and convert_graph.h - working on merging to a single control arg [skip ci]

stuff in scripts for testing prequantized int8 [skip ci]

Add python interface using dict and support dynamic op

debug prints and inference scripts

propagate

First mode where we simply tell the user which ranges are missing if not all are supplied. Still need to infer some ranges backward through shuffle etc. Fix ConvertIdentity bug and add error checking to input binding [skip ci]

cleanup and fix more [skip ci]

Set softmax range. Fix some cases of inferring ranges. Remove theoretical Matmul infer. Don't print missing ranges for internal unnamed tensors

Remove unneeded conversion of dict -> unordered_map (it had bugs with python2 strings causing python2 build to fail) [skip ci]

Fix syntax error and infer backwards in another shuffle situation

Pass weights through quant nodes unchanged. Fix broadcast issue with Relu6. Remove some problematic backwards infers

Cleanup

Cleanup

Change arg to use_calibrate. Add unction to copy ranges. Add comments for features which are incomplete or not used. Fix identity conversion.

Add missing code for Relu6

Add missing range copies and fix issue with const ITensor*

Properly compute symmetric range. Fix const cast bug

Fix bug with use_calibration arg. Change quantization range interface to use Converter class. Change range inferring algorithm to allow for multiple hops. Formatting fixes.

Quant ops are format-agnostic to layout optimizer now. Exclude quant ops from FP16 and FP32 graphs. Rename scales to ranges. Formatting

Add tests for quantization

Fix bug with symmetric range calculation by switching to std::max. Improve tests (they will be broken until Transpose op is added). Add check for when mode is not int8 and calib=true

Fix tests. Add comments and checks, small fixes.

Fix candidate ops check. Log missing tensors unconditionally. Improve some comments. Revert ConvertIdentity to original implementation - the name check in convert_node is sufficient and adding the no-op shuffles introduced problems

Don't use shuffle no-op as conversion for quantize, instead just pass tensor through. Shuffle no-op is not optimized away like we thought so it can interfere with TRT's node fusion, causing reduced accuracy

Const bool

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -726,6 +757,24 @@ class Converter {
+                             bool is_forward) {
+    if (is_forward) {
+      // Carry int8 quantization range forward if available
+      if (quantization_scales_.find(input) != quantization_scales_.end()) {
```
We can do the following to save the extra lookup:
```
auto itr = quantization_scales_.find(input);
if (itr != quantization_scales_.end()) {
  // ...use itr to get the range.
}
```

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -649,6 +662,24 @@ using OpConverter =
 class Converter {
  public:
+  // During conversion, this table is populated with quantization scales per tensor
+  std::unordered_map<nvinfer1::ITensor*, float> quantization_scales_;
```
Please make these private members and add accessors and/or mutators accordingly.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -1314,8 +1364,12 @@ tensorflow::Status ConvertConv2DHelper(
     TFTRT_RETURN_ERROR_IF_NULLPTR(output_tensor, node_def.name());
+    // Store input and output tensors so that quantization ranges from output of shuffle
+    // can be inferred backwards for input
+    ctx.quantization_infer_backwards_[output_tensor] = inp;
```
I think this is equivalent to `ctx.CopyQuantizationRange(inp, output_tensor, false);`?

Also, why not do forward copy here? Actually in general how do we decide when to do forward copy and when to do backward? It seems we can always do both?

trevor-m replied:
Oh yes, it shouldve been CopyQuantizationRange.

Hmm I guess you can always do both. How about renaming CopyQuantizationRange function to something like `MarkQuantizationRangesAsInferrable(input_tensor, output_tensor)`
which adds two edges to the quantization_infer table:
input -> output
output -> input

These edges are then used to propagate ranges across tensors using the algorithm you described here https://github.com/tensorflow/tensorflow/pull/22788#discussion_r223259796

aaroey replied:
That sounds good to me, thanks.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -126,6 +126,14 @@ bool IsTensorRTCandidate(const tensorflow::Node* node) {
     "Min",
+    "Relu6",
+#endif
+#if NV_TENSORRT_MAJOR >= 5
```
What happen if user ask to do FP32/FP16 for a graph with these ops? I know this sounds stupid and normal people shouldn't ask for that, but to make the logic correct, I think the converter should skip handling them instead of removing them. Maybe we can add a precision_mode parameter in this `IsTensorRTCandidate()` function for that.

sreeni-k replied:
+1

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -1813,7 +1991,21 @@ tensorflow::Status ConvertIdentity(
     const std::vector<TRT_TensorOrWeights>& inputs,
     std::vector<TRT_TensorOrWeights>* outputs) {
-  outputs->push_back(inputs.at(0));
+  if (inputs.at(0).is_weights()) {
```
In the PR description you mentioned a bug is fixed with Identity node, may I know what was the bug?

trevor-m replied:
The issue was that if the identity node was the first node in the engine, the input tensor would be renamed from InputPH_0 to something like Identity. This will cause runtime error when the binding for InputPH_0 can't be found. It actually looks like you fixed this in a different way: https://github.com/tensorflow/tensorflow/blob/09c215a9c134dcc52882b153973617e125ad300b/tensorflow/contrib/tensorrt/convert/convert_nodes.cc#L727
However, your solution causes none of the tensors to have their names set - by default TRT names new tensors like "(Unamed Tensor* 1)" etc. but it looks like you check for nullptr or a len of 0.

trevor-m replied:
I should mention that I am using TRT 5.0, the default tensor names may have changed

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -2444,13 +2642,36 @@ tensorflow::Status ConvertMatMulHelper(
+  // for (int64_t i = 0; i < weights.count(); i++)
+  //   if (values[i] > weight_max)
+  //     weight_max = values[i];
+  // LOG(INFO) << "theoretical MatMul range:";
```
Line is too long. Also, please remove or use VLOG.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -2633,6 +2856,54 @@ tensorflow::Status ConvertTopK(Converter& ctx,
 }
 #endif
 
+void ApplyQuantizationRanges(Converter& converter, bool use_calibration) {
```
Please make this a member function of Converter.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -2633,6 +2856,54 @@ tensorflow::Status ConvertTopK(Converter& ctx,
+    }
+  }
+  // Get all tensors from network
+  std::vector<nvinfer1::ITensor*> all_tensors;
```
I believe there will be ~50% duplicates in the vector, so maybe use a std::set instead?

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -2633,6 +2856,54 @@ tensorflow::Status ConvertTopK(Converter& ctx,
+void ApplyQuantizationRanges(Converter& converter, bool use_calibration) {
+  std::vector<nvinfer1::ITensor*> tensors_missing_ranges;
+  // Infer quantization ranges backward through reshapes, maxpool, etc
+  for (auto it : converter.quantization_infer_backwards_) {
```
IIUC, the algorithm here is, use quantization_infer_backwards_[output_tensor] to populate the range information of the input tensor, but only do this once. What if there are multiple hops, e.g. `tensor1->reshape1->tensor2->identity1->tensor3` and we only know the range of `tensor3`? And what if `tensor1` is an input of a different `reshape2` to `tensor4`?

It looks like the right way would be:
```
while (not more range information are added) {
  do infer backwards
  do infer forwards
}
```
what do you think? Thanks.

trevor-m replied:
Hi Lambda, thanks for all of the comments!

Yes that is the correct algorithm. My code assumed a max of one hop which was okay since we only inferred ranges for tensors that we create internally within TF-TRT  (the user could always just provide a quantization node for everything else).

But I agree that we should be smarter about inferring to make this feature easier to use, so I will implement the multi-hop algorithm.

aaroey reviewed (changes requested):
Thanks so much for the PR. A high level comment: I think we need a unit test for this feature. Can we use the mnist example as a unit test?

trevor-m commented on tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -124,8 +124,10 @@ TRTEngineOp::TRTEngineOp(OpKernelConstruction* context)
+  OP_REQUIRES_OK(context,
+                 context->GetAttr("use_calibration", &use_calibration_));
+  calibration_mode_ = (use_calibration_ &&
+      (precision_mode_ == convert::INT8MODE && calibration_data.size() == 0));
```
Yes, in calibration mode the engine will still have calibration_data regardless of any user provided ranges. The workflow for use_calibration=True is the same as it was before, e.g.
1. create_inference_graph() to create calibration graph
2.  send some data through the graph to calibrate
3. calib_graph_to_infer_graph() to create final inference graph

For use_calibration=False, the workflow is simply
1. create_inference_graph() to create final inference graph

wujingyue commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -878,10 +893,11 @@ tensorflow::Status ConvertAfterShapes(ConversionParams& params) {
-    curr_engine.engine_type =
-        (params.is_dyn_op || params.precision_mode == INT8MODE
+    curr_engine.engine_type = ((params.is_dyn_op ||
+      (params.precision_mode == INT8MODE && params.use_calibration))
```
Is precision_mode important? It looks to me that, when use_calibration is true, you should create a dynamic engine regardless of precision_mode.

trevor-m replied:
We are ignoring the use_calibration parameter when precision is not INT8 because calibration doesn't make sense for FP32 or FP16. But if you are talking about in the future when we have INT4 then I can see the issue.

wujingyue replied:
Then should we explicitly check against calibration in fp? E.g. 

```
if (use_calibration and not int8) {
  LOG(FATAL) << "not implemented";
}
engine_type = (is_dyn_op || use_calibration) ? dynamic : static;
```

trevor-m replied:
Good idea, we should do that. In that case, should use_calibration be false by default? That would cause the default INT8 behavior to change, is that okay?

wujingyue replied:
I'm totally fine with calibration off by default. In fact, Waymo doesn't use calibration for now.

aaroey replied:
Hmm I think that will break existing users who uses the INT8 api. Maybe just ignore `use_calibration` when precision_mode is not INT8, and mentioned the behavior in the comment/doc-string of `create_inference_graph()`.

wujingyue replied:
As a general rule of flags, you shouldn't set a flag true by default if it doesn't work with even the default configuration (in this case `precision=fp32). So I believe use_calibration=false by default is the right thing to do. 

IMO, breaking existing users is fine since this is in tf.contrib. Are you aware of any users on int8 calibration? We could notify them about this change.

wujingyue replied:
Btw, not too strong opinion on this. This is a typical trade-off between API stability and future maintenance. @aaroey should make the final call because you are going to maintain this from TF side. 

I'm fine with
```
// This check should appear very early in ConvertGraphDefToTensorRT. 
if (use_calibration and not int8) {
  LOG(ERROR) << "not implemented. fall back to no calibration";
  use_calibration = false;
}

engine_type = (is_dyn_op || use_calibration) ? dynamic : static;
```

wujingyue commented on tensorflow/contrib/tensorrt/ops/trt_engine_op.cc:
```diff
@@ -41,6 +41,7 @@ REGISTER_OP("TRTEngineOp")
     .Attr("workspace_size_bytes: int")
     .Attr("precision_mode: {'FP32', 'FP16', 'INT8', 'INT8CALIB'}")
```
I don't understand why both INT8CALIB and use_calibration. Looks like you need either of them not both.

trevor-m replied:
Not sure why INT8CALIB argument is there either, it appears to be unused. I will remove it unless we decide to switch to an argument like https://github.com/tensorflow/tensorflow/pull/22788#discussion_r223252646

wujingyue commented on tensorflow/contrib/tensorrt/python/trt_convert.py:
```diff
@@ -69,7 +69,8 @@ def tensorrt_rewriter_config(max_batch_size=1,
                              maximum_cached_engines=1,
-                             cached_engine_batch_sizes=None):
+                             cached_engine_batch_sizes=None,
+                             use_calibration=True):
```
I feel not_use_calibration and precision_mode are mutually exclusive. When using calibration, you definitely need precision_mode otherwise you wouldn't figure that out from the graph. When not using calibration, you don't need precision_mode because your graph specifies what precision it needs using QuantizeAndDequantize ops (QuantizeAndDequantize has a num_bits flag). In fact, a user could even do mixed-precision execution in *one* graph by adding some QuantizeAndDequantizes with num_bits=8 and some other QuantizeAndDequantizes with num_bits=4.

aaroey replied:
Thanks, that is a good point. Currently all the TRT engines are using same precision mode and we may want to change that (maybe we can add something to the custom backend API design to allow different settings for different subgraphs).

wujingyue replied:
So I propose not adding INT8_NO_CALIBRATION. precision=int8 and use_calibration=false trigger dynamic range collection and quantization. WDYT, Guangda?

wujingyue replied:
@aaroey

aaroey replied:
Agreed. Closing.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -819,6 +857,39 @@ class Converter {
+            quantization_scales_.end()) {
+          it = quantization_infer_.erase(it);
+        }
+        else {
```
Thanks for the changes here.
nit: `} else {`.

aaroey commented on tensorflow/contrib/tensorrt/python/trt_convert.py:
```diff
@@ -168,6 +176,11 @@ def create_inference_graph(input_graph_def,
       use this list to determine the batch sizes of the cached engines, instead
       of making the decision on the fly. This is useful when we know the most
       common batch size(s) the application is going to generate.
+    use_calibration: selects the behavior when precision_mode is INT8.
```
Please make it clear that if precision_mode is not INT8 this option will be ignored.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -726,6 +739,69 @@ class Converter {
+        if (std::string(tensor->getName()).find("(Unnamed ITensor* ") ==
+            std::string::npos) {
+          LOG(WARNING) << "Quantization range was not found for "
+            << tensor->getName()
```
nit: please fix the alignment.

aaroey commented on tensorflow/contrib/tensorrt/python/trt_convert.py:
```diff
@@ -137,6 +144,7 @@ def create_inference_graph(input_graph_def,
                            is_dynamic_op=False,
                            maximum_cached_engines=1,
                            cached_engine_batch_sizes=None,
+                           use_calibration=True,
```
Please add a test, maybe using the mnist example.

wujingyue replied:
+1

trevor-m replied:
I've added some tests. I also have an example script which can be used to train and test a simple model on MNIST - should I include that as well? Where would it go?

wujingyue commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -126,6 +126,14 @@ bool IsTensorRTCandidate(const tensorflow::Node* node) {
+    "Relu6",
+#endif
+#if NV_TENSORRT_MAJOR >= 5
+    "QuantizeV2",
```
FYI, I don't think we will ever have QuantizeV2 and Dequantize in the input graph. Of course, doesn't hurt to support them.

trevor-m replied:
Are these ops no longer supported in favor of QuantizeAndDequantizeV2/3? I noticed there is no python API to add QuantizeAndDequantizeV2/3 nodes.

sreeni-k replied:
Though our initial example used the QunatizeV2/Dequantize ops, going forward our models will use only QuantizeAndDequantizeV2/3. The last example already uses the new combined op.

trevor-m commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -760,6 +837,17 @@ class Converter {
+  // During conversion, this table is populated with quantization scales per
+  // tensor. ApplyQuantizationRanges() will use this table to set the TensoRT
+  // quantization ranges.
+  std::unordered_map<nvinfer1::ITensor*, float> quantization_scales_;
```
I see, it seems I am meaning the floating point range. I will update the language to use "ranges" instead of scales.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -1577,6 +1712,104 @@ tensorflow::Status ConvertActivation(
+  nvinfer1::ITensor* output_tensor = layer->getOutput(0);
+
+  // Store ranges for input and output of dummy op
+  ctx.ProvideQuantizationRange(tensor, min_range, max_range);
```
Here we assume that `tensor` is not used elsewhere as input to other nodes. If user has a crazy graph where `tensor` is fed into different ops and some but not all of them are QDQV2, or the ranges of the QDQV2s are different, there will be problem. I don't see any easy way to detect this with the current implementation, so may be add comment and a `LOG(WARNING)`?

Do you have a better idea? Also @wujingyue

wujingyue replied:
You would need mixed-precision to solve these cases. One potential solution is to keep the tensor in fp and quantize it on one path bot not the other, or quantize it differently on both paths. 

Given TensorRT doesn't plan to support mixed-precision in short term, I'd emit a fatal error if this happens. We can make sure quantized training doesn't produce graphs like that at the first place.

aaroey commented on tensorflow/core/grappler/optimizers/layout_optimizer.cc:
```diff
@@ -108,6 +108,7 @@ std::set<string> GetOpsFormatAgnostic() {
                                           "Conj",
                                           "Cos",
                                           "Cosh",
+                                          "Dequantize",
```
@rmlarsen for the changes in layout_optimizer.cc

aaroey commented on tensorflow/contrib/tensorrt/test/tf_trt_integration_test_base.py:
```diff
@@ -514,9 +520,20 @@ def _Test(self):
+        dynamic_engine=dynamic_engine,
+        test_name=test_name,
+        use_calibration=use_calibration)
+    if NeedsCalibration(run_params):
```
Maybe add:
```
if precision_mode != 'INT8' and use_calibration:
  continue
```
So we don't run tests for FP32/FP16 twice.

aaroey replied:
Also based on what you said, if `precision_mode == 'INT8' and not use_calibration`, engine building should fail, meaning that all existing tests will fail, is that right?

If so we may want to let `ShouldRunTest()` return `False` if `precision_mode == 'INT8' and not use_calibration` by default, and override it in your newly added tests (which you already did).

trevor-m replied:
Oh right, that makes sense. I will also revert `NeedsCalibration()` back to `IsQuantizationMode()` because both methods of using INT8 require special considerations.

trevor-m replied:
> Also based on what you said, if `precision_mode == 'INT8' and not use_calibration`, engine building should fail, meaning that all existing tests will fail, is that right?
> If so we may want to let `ShouldRunTest()` return `False` if `precision_mode == 'INT8' and not use_calibration` by default, and override it in your newly added tests (which you already did).

I will also have to modify every test that overrides `ShouldRunTest()` to exclude INT8 w/o calibration, correct?

trevor-m commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -712,7 +712,7 @@ class Converter {
       if (output.is_tensor()) {
         const char* tensor_name = output.tensor()->getName();
-        if (tensor_name == nullptr || std::strlen(tensor_name) == 0) {
+        if (std::string(tensor_name).find(kInputPHName) == std::string::npos) {
```
The old way ends up never setting the names for any tensors - TRT initializes tensor names to "(Unnamed ITensor* N)" so the condition `(tensor_name == nullptr || std::strlen(tensor_name) == 0)` can never be met.

We now rely on the tensor names for reporting which tensors are missing quantization ranges.

aaroey replied:
Thanks, I think I missed that. Would you please add comment for this tensor name initialization behavior?

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -794,9 +794,9 @@ class Converter {
-            << tensor->getName()
-            << ". This might be okay if TensorRT does not need the range"
-            << "(e.g. due to node fusion).";
+                       << tensor->getName() << ". "
```
What will happen if after fusing there are still missing ranges? In this case IIUC the engine op will not have a null calibrator and will it crash? Please comment.

trevor-m replied:
If there are still missing ranges, Builder::buildCudaEngine will return nullptr. Which we catch in `ConvertGraphDefToEngine` and return `tensorflow::errors::Internal("Failed to build TensorRT engine")`.

aaroey replied:
I see, thanks. Please document this behavior in both comment and the LOG message.

aaroey commented on tensorflow/contrib/tensorrt/test/quantization_test.py:
```diff
@@ -0,0 +1,162 @@
+        expected_output_dims=[(100, 10)])
+
+  def ShouldRunTest(self, run_params):
+    return (run_params.precision_mode == "INT8" and
```
Will it crash at op execution time when `run_params.use_calibration == False` in this case?

trevor-m replied:
Oh right, I must've gotten it mixed up.

aaroey commented on tensorflow/contrib/tensorrt/test/tf_trt_integration_test_base.py:
```diff
@@ -514,18 +525,24 @@ def _Test(self):
         # supported yet.
         continue
-      if not dynamic_engine:
+      if not dynamic_engine and use_calibration:
```
Why changing this line? IIRC, the original condition is "run INT8 conversion if it's dynamic engine", the new one is "run INT8 conversion if it's dynamic engine or if it's static engine without calibration". I think we don't want to run conversion with static engine which is not supported yet.

trevor-m replied:
Oh right, thanks for catching that! My intention here was to run the following:
int8=1, dynamic, use_calibration=1
int8=1, dynamic, use_calibration=0
~~int8=1, static, use_calibration=1~~ <- exclude this
int8=1, static, use_calibration=0

So the correct logic should be
`if not dynamic_engine or not use_calibration`
Does that seem right?

aaroey replied:
INT8+static engine will fall back to dynamic engine, see [here](https://github.com/tensorflow/tensorflow/pull/22788/files#diff-63b6c12e8d9c0da3f1a4476c60494ea5R910), is that what you want to test?

trevor-m replied:
>```curr_engine.engine_type = ((params.is_dyn_op || params.use_calibration)
             ? EngineInfo::EngineType::TRTDynamic
             : EngineInfo::EngineType::TRTStatic);```

use_calibration+static engine should fall back to dynamic engine, but not use_calibration+static engine will be static

aaroey replied:
I see, make sense. Thanks.

trevor-m replied:
Thanks, sorry for the confusion

aaroey replied:
I read this one more time and is confused again until read through the discussions here. So please add a comment saying the fact that, !use_calibration+static engine will be static, and we want to test that.

aaroey reviewed (changes requested):
@trevor-m Thanks so much for the fixes. Please help to fix the remaining comments, and after #22874 is merged this one is good to go.

trevor-m commented:
Hi Lambda, I'm going to start rebasing this PR. Afterwards I will work on improving the tests and adding some C++ unit tests as well.

trevor-m committed:
Add FakeQuantWithMinMaxArgs

trevor-m committed:
Add more checks on quantize node inputs.

Add ConvertQuantize tests

Fix line length formatting

Set quantization range for IConstantLayer created when weights are reshaped via PrepareTensorForShape. Since the weights are constant we can use the value of the weights to determine the range.

Add missing inference of quantization range. Allow ConvertReshape to work on weights

Fix reshape for weights

Fix reshape for weights again

trevor-m committed:
Fix merge conflicts

trevor-m committed:
Improve use_calibration arg documentation

trevor-m commented:
Hi Lambda, the PR is ready to go now after you review the latest changes.

I have a test script to train or evaluate a simple model on MNIST using this feature. Would it make sense to include it in the test suite here? How do you handle tests which use external data (the model checkpoint in this case).

aaroey commented:
@trevor-m thanks for making the changes, I'll take a look soon. Please add the mnist test case, it's fine to include test data as long as it's not too big.

trevor-m committed:
Quantization-aware training full test on MNIST

trevor-m committed:
Add quantization mnist test to BUILD

aaroey commented on tensorflow/contrib/tensorrt/test/quantization_mnist_test.py:
```diff
@@ -0,0 +1,197 @@
+        use_trt=False,
+        batch_size=128,
+        num_epochs=None,
+        model_dir='./quantization_mnist_test_data')['accuracy']
```
Please add a build rule for the test data and access it maybe using `test.test_src_dir_path()`. There is an example [BUILD](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/BUILD#L3021) and [py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/self_adjoint_eig_op_test.py#L78) file you can follow.

aaroey commented on tensorflow/contrib/tensorrt/test/quantization_mnist_test.py:
```diff
@@ -0,0 +1,197 @@
+import tensorflow as tf
```
Please add copyright header.

trevor-m commented:
Thanks for reviewing! Perhaps we should then remove Dequantize as well - it won't do much by itself and it requires inputs of qint8.

trevor-m commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -1004,6 +1030,21 @@ Status Converter::PrepareTensorForShape(const TRT_TensorOrWeights& input,
     *tensor = layer->getOutput(0);
+    // We need to set a quantization range for the output tensor of the
+    // IConstantLayer. Here we set the range to [min(weights), max(weights)].
+    float min_range = 0.0f;
```
I agree that with use_calibration=True we should not set a range here. 
When precision_mode is not int8, the quantization ranges will be ignored by TRT. Are you concerned about wasting CPU cycles computing the min/max in this case?

aaroey replied:
No I'm more concerning about maintenance, I think we should make the condition clear so later when people look at the code they don't get confused.

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_nodes.cc:
```diff
@@ -95,6 +95,9 @@ inline tensorflow::Status ConvertDType(tensorflow::DataType tf_dtype,
     case tensorflow::DataType::DT_INT8:
       *trt_dtype = nvinfer1::DataType::kINT8;
       break;
+    case tensorflow::DataType::DT_QINT8:
```
No they're not the same:

- DT_INT8 is not quantize type, it doesn't require min/max range information for computation (similar to DT_INT32, DT_FLOAT32, etc)
- DT_QINT8 is quantize type and we do need range information for any computation. I think this is equivalent to TRT's INT8, since TRT's INT8 needs either calibration to populate the ranges or use setDynamicRange(). This is why I think converting DT_INT8 to TRT's INT8 may be wrong.

trevor-m commented:
> @trevor-m thanks for making the changes, I'll take a look soon. Please add the mnist test case, it's fine to include test data as long as it's not too big.

@aaroey 
I have implemented all of your requested changes and I am just finishing up the tests. However I am getting this error with the MNIST test. Have you encountered this before? I suspect it has to do with how bazel is executing the test...
```
Node construction failed withInvalid argument: Op type not registered 'TRTEngineOp' in binary running on 8eb481f0e4e9. Make sure the Op and Kernel are registered in the binary running in this process. Note that if you are loading a saved graph which used ops from tf.contrib, accessing (e.g.) `tf.contrib.resampler` should be done before importing the graph, as contrib ops are lazily registered when the module is first accessed. while building NodeDef 'my_trt_op_0'
2018-11-09 00:41:44.474018: W tensorflow/contrib/tensorrt/convert/convert_graph.cc:1017] Engine my_trt_op_0 creation for segment 0, composed of 29 nodes failed: Invalid argument: Op type not registered 'TRTEngineOp' in binary running on 8eb481f0e4e9. Make sure the Op and Kernel are registered in the binary running in this process. Note that if you are loading a saved graph which used ops from tf.contrib, accessing (e.g.) `tf.contrib.resampler` should be done before importing the graph, as contrib ops are lazily registered when the module is first accessed. while building NodeDef 'my_trt_op_0'. Skipping...
```

aaroey commented:
Thanks @trevor-m for the changes! If the test failure is the only problem left, please upload the commits, I can help to take a look.

trevor-m committed:
Fixes for @aaroey's review.

Fix MNIST integration test imports and add the data dependencies as a build rule.
GetWeightRange will now only be called in int8mode no calib mode.
Make precision_mode a member of TrtCandidateSelect instead of argument to IsCandidate.
Make precision_mode, use_calibration members of Converter
Remove QINT8 from ConvertDType. Remove QINT8, UINT8, INT8, QUINT8 from RedorderCK. These are no longer needed after removing QuantizeV2 and Dequantize support.
Refactor ApplyQuantizatioNRanges() to make it easier to test.
ConvertQuantize will explicitly look for the different ops by name.
Add checks to Relu6.
Add unit tests for new functions in Converter.
Add comments to clarify some logic

aaroey reviewed (dismissed):
Thanks a lot for adding the tests! I'll handle the rest.

trevor-m commented:
Cool, thanks for all of the helpful feedback!

aaroey committed:
Fix some formatting/compilation issues

trevor-m committed:
Fix data dependency bug

trevor-m commented:
Sorry it looks like it somehow automatically requested a lot of people to review - please ignore.

trevor-m committed:
Fix QuantizationMissingAllRangesTest: the network was too simple and TRT was chosing not to quantize it.

trevor-m committed:
Identified a few more situations where quantization range needs to be inferred. In some cases, it is impossible to infer these ranges.

aaroey commented:
Hi @trevor-m, thanks for the fix. At this point, since the PR is already in the merging state internally, it's a bit hard to merge your new changes. I can try to merge them, just FYI. If you have new changes, would you please issue a separate PR after this one is merged? Thanks.

trevor-m commented:
> Hi @trevor-m, thanks for the fix. At this point, since the PR is already in the merging state internally, it's a bit hard to merge your new changes. I can try to merge them, just FYI. If you have new changes, would you please issue a separate PR after this one is merged? Thanks.

Oh I see, sorry about that. If you are unable to merge any of the changes let me know which ones and I can issue a separate PR later. There won't be any other changes from now on.

aaroey commented:
@trevor-m not a problem at all. It's my fault didn't mention that. I'll let you know how it goes later. :)

aaroey commented:
This PR is actually merged. Not sure why it's still open like this.

trevor-m commented:
> This PR is actually merged. Not sure why it's still open like this.

Thanks Lambda! I probably messed it up when I pushed those commits after you had started merging.

aaroey committed:
Merge with upstream master

aaroey committed:
Merge with upstream again

aaroey commented:
@trevor-m no worries I'll try again.
